### ISSUE TITLE
NullPointerException when trying to send message using ServiceBusSenderAsyncClient

### ISSUE THREAD

**Issue by c002521** [Azure/azure-sdk-for-java#19281]
**Query/Question**
Getting NullPointerException when trying to send message using ServiceBusSenderAsyncClient. This is not reproducible in local, however this happens frequently in production after few messages are processed successfully.
This is a Java 11 + Spring Boot (2.3.8.RELEASE) application with azure-messaging-servicebus:7.0.2 dependency.

```
@Bean(name = "topicClient")
public ServiceBusSenderAsyncClient topicClient() {
    return new ServiceBusClientBuilder()
            .connectionString(serviceBusConfigProperties.getConnectionString()).sender()
            .topicName(serviceBusConfigProperties.getOrderStatusTopicName()).buildAsyncClient();
}
```

```
topicClient
        .sendMessage(message)
        .subscribe(
                sentSignal -> LOGGER.debug("Sent Message to Topic"),
                errorSignal -> handleFailedMessages(message, errorSignal));

```
Send message on subscribe receives error signal and the messages are not reaching the topic.
handleFailedMessages prints the exception stacktrace. The ServiceBusFailureReason is GENERAL_ERROR.

```
com.azure.messaging.servicebus.ServiceBusException
	at com.azure.messaging.servicebus.ServiceBusSenderAsyncClient.mapError(ServiceBusSenderAsyncClient.java:737)
	at reactor.core.publisher.Mono.lambda$onErrorMap$30(Mono.java:3325)
	at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:88)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:165)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onError(FluxPeekFuseable.java:227)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:185)
	at reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:251)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:165)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onError(FluxMapFuseable.java:134)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onError(FluxMapFuseable.java:134)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:113)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1787)
	at reactor.core.publisher.MonoCallable.subscribe(MonoCallable.java:61)
	at reactor.core.publisher.InternalMonoOperator.subscribe(InternalMonoOperator.java:64)
	at reactor.core.publisher.MonoFlatMap$FlatMapMain.onNext(MonoFlatMap.java:150)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1787)
	at com.azure.core.amqp.implementation.AmqpChannelProcessor.subscribe(AmqpChannelProcessor.java:229)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4252)
	at reactor.core.publisher.Mono.subscribeWith(Mono.java:4363)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4223)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4159)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4131)
         : Removed lines for readability
Caused by: java.lang.NullPointerException
	at org.apache.qpid.proton.engine.impl.ConnectionImpl.session(ConnectionImpl.java:91)
	at org.apache.qpid.proton.engine.impl.ConnectionImpl.session(ConnectionImpl.java:39)
	at com.azure.core.amqp.implementation.ReactorConnection.lambda$createSession$6(ReactorConnection.java:195)
	at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
	at com.azure.core.amqp.implementation.ReactorConnection.lambda$createSession$7(ReactorConnection.java:192)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:107)
	... 165 more

```

***Why is this not a Bug or a feature Request?***
Unable to reproduce in local

**Setup (please complete the following information if applicable):**
 - Version of the Library used
 * Java 11
 * Spring Boot (2.3.8.RELEASE)
 * azure-messaging-servicebus 7.0.2
 
 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [ ] Query Added
- [ ] Setup information Added

**Comment by joshfree** [comment]
Thanks for reporting this issue, @c002521.  @hemanttanwar can you please investigate for the upcoming March release?

**Comment by c002521** [comment]
@joshfree  @hemanttanwar 
On further investigating, I was able to reproduce this in my local.
The issue is with the AMQP connection. After exceeding idle time, the connection/session is closed. However it is unable to create new connection. 
More exceptions observed in the logs before ServiceBusException are shared below.
```
2021-02-18 08:10:19,416 INFO  [reactor-executor-1] SendLinkHandler: onLinkRemoteClose connectionId[MF_24e5f5_XXXXXXXX], linkName[XXXXXXXX-topic], errorCondition[amqp:link:detach-forced], errorDescription[Idle link tracker, link XXXXXXXX-topic has been idle for 1800000ms TrackingId:XXXXXXXX_G27, SystemTracker:client-link20054939, Timestamp:2021-02-18T14:10:19]
2021-02-18 08:10:19,417 INFO  [reactor-executor-1] SendLinkHandler: processOnClose connectionId[MF_24e5f5_XXXXXXXX], linkName[XXXXXXXX-topic], errorCondition[amqp:link:detach-forced], errorDescription[Idle link tracker, link XXXXXXXX-topic has been idle for 1800000ms TrackingId:XXXXXXXX_G27, SystemTracker:client-link20054939, Timestamp:2021-02-18T14:10:19]
2021-02-18 08:10:19,417 DEBUG [reactor-executor-1] ReactorSender: connectionId[MF_24e5f5_XXXXXXXX], path[XXXXXXXX-topic], linkName[XXXXXXXX-topic]: State CLOSED
2021-02-18 08:10:19,417 INFO  [reactor-executor-1] ReactorSession: linkName[XXXXXXXX-topic]: Error occurred. Removing and disposing send link.
com.azure.core.amqp.exception.AmqpException: Idle link tracker, link XXXXXXXX-topic has been idle for 1800000ms TrackingId:XXXXXXXX_G27, SystemTracker:client-link20054939, Timestamp:2021-02-18T14:10:19, errorContext[NAMESPACE: XXXXXXX.servicebus.windows.net, PATH: XXXXXXXX-topic, REFERENCE_ID: XXXXXXXX0_G27, LINK_CREDIT: 297]
at com.azure.core.amqp.implementation.ExceptionUtil.toException(ExceptionUtil.java:85)
at com.azure.core.amqp.implementation.handler.LinkHandler.handleRemoteLinkClosed(LinkHandler.java:101)
at com.azure.core.amqp.implementation.handler.LinkHandler.onLinkRemoteClose(LinkHandler.java:50)
at com.azure.core.amqp.implementation.handler.SendLinkHandler.onLinkRemoteClose(SendLinkHandler.java:20)
at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:176)
at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
at com.azure.core.amqp.implementation.ReactorExecutor.run(ReactorExecutor.java:82)
at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
at java.base/java.lang.Thread.run(Thread.java:834)
2021-02-18 08:10:19,418 INFO  [reactor-executor-1] SendLinkHandler: onLinkLocalClose connectionId[MF_24e5f5_XXXXXXXX], linkName[XXXXXXXX-topic], errorCondition[amqp:link:detach-forced], errorDescription[Idle link tracker, link XXXXXXXX-topic has been idle for 1800000ms TrackingId:XXXXXXXX_G27, SystemTracker:client-link20054939, Timestamp:2021-02-18T14:10:19]
:
2021-02-18 08:15:19,676 DEBUG [reactor-executor-1] RequestResponseChannel: Attempted 1 times to get a new AMQP connection
2021-02-18 08:15:19,676 WARN  [reactor-executor-1] RequestResponseChannel: Retry #1. Transient error occurred. Retrying after 4511 ms.
The connection was inactive for more than the allowed 300000 milliseconds and is closed by container 'LinkTracker'. TrackingId:XXXXXXXX0_G27, SystemTracker:gateway7, Timestamp:2021-02-18T14:15:19, errorContext[NAMESPACE: XXXXXXX.servicebus.windows.net, PATH: $cbs, REFERENCE_ID: cbs:sender, LINK_CREDIT: 96]
com.azure.core.amqp.exception.AmqpException: The connection was inactive for more than the allowed 300000 milliseconds and is closed by container 'LinkTracker'. TrackingId:XXXXXXXX0_G27, SystemTracker:gateway7, Timestamp:2021-02-18T14:15:19, errorContext[NAMESPACE: XXXXXXX.servicebus.windows.net, PATH: $cbs, REFERENCE_ID: cbs:sender, LINK_CREDIT: 96]
at com.azure.core.amqp.implementation.ExceptionUtil.toException(ExceptionUtil.java:85)
at com.azure.core.amqp.implementation.handler.LinkHandler.handleRemoteLinkClosed(LinkHandler.java:101)
at com.azure.core.amqp.implementation.handler.LinkHandler.onLinkRemoteClose(LinkHandler.java:50)
at com.azure.core.amqp.implementation.handler.SendLinkHandler.onLinkRemoteClose(SendLinkHandler.java:20)
at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:176)
at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
at com.azure.core.amqp.implementation.ReactorExecutor.run(ReactorExecutor.java:82)
at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
at java.base/java.lang.Thread.run(Thread.java:834)
2021-02-18 08:15:19,678 ERROR [reactor-executor-1] RequestResponseChannel: cbs - Exception in RequestResponse links. Disposing and clearing unconfirmed sends.
The connection was inactive for more than the allowed 300000 milliseconds and is closed by container 'LinkTracker'. TrackingId:XXXXXXXX0_G27, SystemTracker:gateway7, Timestamp:2021-02-18T14:15:19, errorContext[NAMESPACE: XXXXXXX.servicebus.windows.net, PATH: $cbs, REFERENCE_ID: cbs:sender, LINK_CREDIT: 96]
com.azure.core.amqp.exception.AmqpException: The connection was inactive for more than the allowed 300000 milliseconds and is closed by container 'LinkTracker'. TrackingId:XXXXXXXX0_G27, SystemTracker:gateway7, Timestamp:2021-02-18T14:15:19, errorContext[NAMESPACE: XXXXXXX.servicebus.windows.net, PATH: $cbs, REFERENCE_ID: cbs:sender, LINK_CREDIT: 96]
at com.azure.core.amqp.implementation.ExceptionUtil.toException(ExceptionUtil.java:85)
at com.azure.core.amqp.implementation.handler.LinkHandler.handleRemoteLinkClosed(LinkHandler.java:101)
at com.azure.core.amqp.implementation.handler.LinkHandler.onLinkRemoteClose(LinkHandler.java:50)
at com.azure.core.amqp.implementation.handler.SendLinkHandler.onLinkRemoteClose(SendLinkHandler.java:20)
at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:176)
at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
at com.azure.core.amqp.implementation.ReactorExecutor.run(ReactorExecutor.java:82)
at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
at java.base/java.lang.Thread.run(Thread.java:834)
:
2021-02-18 08:15:19,678 INFO  [reactor-executor-1] ReceiveLinkHandler: processOnClose connectionId[MF_24e5f5_XXXXXXXX], linkName[cbs:receiver], errorCondition[amqp:connection:forced], errorDescription[The connection was inactive for more than the allowed 300000 milliseconds and is closed by container 'LinkTracker'. TrackingId:XXXXXXXX0_G27, SystemTracker:gateway7, Timestamp:2021-02-18T14:15:19]
2021-02-18 08:15:19,678 INFO  [reactor-executor-1] SessionHandler: onSessionRemoteClose connectionId[cbs-session], entityName[MF_24e5f5_XXXXXXXX], condition[Error{condition=null, description='null', info=null}]
2021-02-18 08:15:19,678 INFO  [reactor-executor-1] SessionHandler: onSessionRemoteClose closing a local session for connectionId[MF_24e5f5_XXXXXXXX], entityName[cbs-session], condition[null], description[null]
2021-02-18 08:15:19,678 DEBUG [reactor-executor-1] ReactorSession: connectionId[MF_24e5f5_XXXXXXXX], sessionName[cbs-session], state[CLOSED]
2021-02-18 08:15:19,679 INFO  [reactor-executor-1] ReactorConnection: connectionId[MF_24e5f5_XXXXXXXX] sessionName[cbs-session]: Error occurred. Removing and disposing session.
com.azure.core.amqp.exception.AmqpException: onSessionRemoteClose connectionId[MF_24e5f5_XXXXXXXX], entityName[cbs-session], condition[Error{condition=null, description='null', info=null}], errorContext[NAMESPACE: XXXXXXX.servicebus.windows.net, PATH: cbs-session]
at com.azure.core.amqp.implementation.handler.SessionHandler.onSessionRemoteClose(SessionHandler.java:130)
at org.apache.qpid.proton.engine.BaseHandler.handle(BaseHandler.java:152)
at org.apache.qpid.proton.engine.impl.EventImpl.dispatch(EventImpl.java:108)
at org.apache.qpid.proton.reactor.impl.ReactorImpl.dispatch(ReactorImpl.java:324)
at org.apache.qpid.proton.reactor.impl.ReactorImpl.process(ReactorImpl.java:291)
at com.azure.core.amqp.implementation.ReactorExecutor.run(ReactorExecutor.java:82)
at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
at java.base/java.lang.Thread.run(Thread.java:834)
:
2021-02-18 08:15:24,189 INFO  [parallel-5] RequestResponseChannel: Retry #1. Requesting from upstream.
2021-02-18 08:15:24,190 INFO  [parallel-5] RequestResponseChannel: namespace[MF_24e5f5_XXXXXXXX] entityPath[$cbs]: Connection not requested, yet. Requesting one.
2021-02-18 08:15:24,191 WARN  [parallel-5] RequestResponseChannel: Non-retryable error occurred in connection.
java.lang.NullPointerException: null
at org.apache.qpid.proton.engine.impl.ConnectionImpl.session(ConnectionImpl.java:91)
at org.apache.qpid.proton.engine.impl.ConnectionImpl.session(ConnectionImpl.java:39)
at com.azure.core.amqp.implementation.ReactorConnection.lambda$createSession$6(ReactorConnection.java:195)
at java.base/java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1705)
at com.azure.core.amqp.implementation.ReactorConnection.lambda$createSession$7(ReactorConnection.java:192)
at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:107)
at reactor.core.publisher.Operators$MonoSubscriber.request(Operators.java:1877)
at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:162)
at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:162)
at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:162)
at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.request(FluxPeekFuseable.java:137)
at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.request(Operators.java:2129)
at com.azure.core.amqp.implementation.AmqpChannelProcessor.requestUpstream(AmqpChannelProcessor.java:286)
at com.azure.core.amqp.implementation.AmqpChannelProcessor.lambda$onError$4(AmqpChannelProcessor.java:172)
at reactor.core.publisher.LambdaMonoSubscriber.onNext(LambdaMonoSubscriber.java:168)
at reactor.core.publisher.MonoDelay$MonoDelayRunnable.run(MonoDelay.java:117)
at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:68)
at reactor.core.scheduler.SchedulerTask.call(SchedulerTask.java:28)
at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
at java.base/java.lang.Thread.run(Thread.java:834)
2021-02-18 08:15:24,191 INFO  [parallel-5] RequestResponseChannel: namespace[MF_24e5f5_XXXXXXXX] entityPath[$cbs]: Error in AMQP channel processor. Notifying 0 subscribers.

```

**Comment by hemanttanwar** [comment]
@c002521 , I am trying to replicate this issue while using async sender. But I do not see this problem.

Here is link to this application which setup spring boot application. This does not run forever, but it shows how to set up client.

https://github.com/Azure/azure-sdk-for-java/tree/master/sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-servicebus

The AsyncSender link after hours of idle time when I am testing spring boot application.  I will upload that application in some time  for you to test as well.

**Comment by c002521** [comment]
@hemanttanwar Please give access to Demo Application.

**Comment by hemanttanwar** [comment]
@c002521  Sorry I missed to make it public, it public now. So you can access it.

**Comment by c002521** [comment]
Thanks @hemanttanwar 

I am able to reproduce this issue with my code in the local. Although it is not consistent.
Before invoking the sendMessage, we are performing some DB operations.
 e.g. Rest API/Controller -> Service (DB operations) -> ServiceBusSenderAsyncClient.sendMessage().subscribe()  (We are calling subscribe and not block).

My pom have other dependencies which I included in demo application code <java-coffee-order-dependency-tree attached>.
However I am unable to reproduce this issue with Demo application code.

btw, similar issue is reported on other site: https://www.gitmemory.com/issue/Azure/azure-sdk-for-java/18293/759843692

[java-coffee-order-dependency-tree.txt](https://github.com/Azure/azure-sdk-for-java/files/6094175/java-coffee-order-dependency-tree.txt)

**Comment by c002521** [comment]
@hemanttanwar 
Issue is reproducible with 'New Relic' agent.

I modified the [CoffeeShopOrderController.java](https://github.com/hemanttanwar/java.coffee.order/blob/main/src/main/java/com/azure/sample/messaging/order/controller/CoffeeShopOrderController.java) to set the SessionId & MessageId as the topic is enabled for Partitioning and Duplicate detection.

Run Demo application with below settings.
- Update application.properties with connection string and queue/topic name.
- JVM args: -javaagent:path/to/newrelic.jar
- Environment variables: NEW_RELIC_LICENSE_KEY=<Your Key>;NEW_RELIC_APP_NAME=java-coffee-order

Invoke below http endpoint after server startup and wait for 30-35 mins, before invoking the endpoint second time.
curl -X POST -d "order=1" http://127.0.0.1:8080/sendOrder

p.s: 
([New Relic java agent jar](http://download.newrelic.com/newrelic/java-agent/newrelic-agent/current/newrelic.jar))

**Comment by conniey** [comment]
I've seen this issue while I worked on reliably reconnecting an AMQP connection. This should be resolved in: [7.2.0](https://search.maven.org/artifact/com.azure/azure-messaging-servicebus/7.2.0/jar) because we did a major refactor of azure-core-amqp.
